### PR TITLE
Implement type definition syntax

### PR DESCRIPTION
**Pull request by seanyoung**
Ethereum Solidity introduces a new syntax for user defined types, see:

https://docs.soliditylang.org/en/v0.8.13/types.html

Documentation for the syntax is in the documentation of this commit, and not repeated here.

Signed-off-by: Sean Young <sean@mess.org>

### CODE REVIEW

File: src/sema/mod.rs [hunk 1]
```diff
@@ -478,6 +451,14 @@ impl ast::Namespace {
...
                     ));
                 }
+                ast::Symbol::Type(loc, _) => {
+                    self.diagnostics.push(ast::Diagnostic::error_with_note(
+                        id.loc,
+                        format!("{} is already defined as a type", id.name),
+                        *loc,
+                        "location of previous definition".to_string(),
+                    ));
+                }
```
Comment (LucasSte) [inline src/sema/mod.rs:461]: Your tests do not cover this change

File: src/sema/mod.rs [hunk 2]
```diff
@@ -560,6 +541,14 @@ impl ast::Namespace {
...
                         ));
                     }
+                    ast::Symbol::Type(loc, _) => {
+                        self.diagnostics.push(ast::Diagnostic::warning_with_note(
+                            id.loc,
+                            format!("{} is already defined as a type", id.name),
+                            *loc,
+                            "location of previous definition".to_string(),
+                        ));
+                    }
```
Comment (LucasSte) [inline src/sema/mod.rs:551]: This is not covered too.

File: src/sema/mod.rs [hunk 3]
```diff
@@ -978,6 +970,15 @@ impl ast::Namespace {
...
             }
+            Some(ast::Symbol::Type(loc, _)) => {
+                let loc = *loc;
+                self.diagnostics.push(ast::Diagnostic::warning_with_note(
+                    id.loc,
+                    format!("declaration of ‘{}’ shadows type", id.name),
+                    loc,
+                    "previous declaration of type".to_string(),
+                ));
+            }
```
Comment (LucasSte) [inline src/sema/mod.rs:981]: This one is not covered too

File: src/sema/mod.rs [hunk 4]
```diff
@@ -1408,6 +1410,13 @@ impl ast::Namespace {
                     ));
                     return Err(());
                 }
+                Some(ast::Symbol::Type(..)) => {
+                    diagnostics.push(ast::Diagnostic::decl_error(
+                        id.loc,
+                        format!("‘{}’ is a type", id.name),
+                    ));
+                    return Err(());
+                }
```
Comment (LucasSte) [inline src/sema/mod.rs:1419]: Neither this one

File: src/sema/mod.rs [hunk 5]
```diff
@@ -761,6 +750,9 @@ impl ast::Namespace {
             Some(ast::Symbol::Import(..)) => {
                 ast::Diagnostic::decl_error(id.loc, format!("`{}' is an import", id.name))
             }
+            Some(ast::Symbol::Type(..)) => {
+                ast::Diagnostic::decl_error(id.loc, format!("`{}' is a type", id.name))
+            }
```
Comment (LucasSte) [inline src/sema/mod.rs:755]: This one is not covered as well

**Review by LucasSte — CHANGES_REQUESTED** [review]
Your tests

File: src/sema/mod.rs (hunk 1 above)
Comment (seanyoung) [reply src/sema/mod.rs:461]: fixed

File: src/sema/mod.rs (hunk 2 above)
Comment (seanyoung) [reply src/sema/mod.rs:551]: fixed

File: src/sema/mod.rs (hunk 4 above)
Comment (seanyoung) [reply src/sema/mod.rs:1419]: fixed

File: src/sema/mod.rs (hunk 5 above)
Comment (seanyoung) [reply src/sema/mod.rs:755]: Fixed

File: src/sema/mod.rs (hunk 3 above)
Comment (seanyoung) [reply src/sema/mod.rs:975]: fixed

**Review by LucasSte — APPROVED** [review]

**Pull request merged**